### DISCUSSION TITLE
Validation of user input

### DISCUSSION THREAD

**Discussion by StefanMunnes** [surveydown-dev/surveydown#125] [Feature Requests]
### Description

I think it would be helpful if one could define some checks for the input and give a note if it doesn't match the expectations. I saw there are some packages out there, but just tested [shinyvalidate ](https://github.com/rstudio/shinyvalidate). I will stick to the important example of valid e-mail address input. For now, I was able to implement the following code in the APP.R server function:
```r
  iv <- InputValidator$new()
  iv$add_rule("mail", sv_email("No valid email address."))
  iv$add_rule("mail, sv_regex("^.+?@.+?\..{2,}$", "No valid email address.")) # general alternative
  iv$enable()
```
This package offers much more helper functions (e.g. sv_equal or sv_between) and the error messages fit quite well in the output. I was wondering if you are interested in implementing this functionality in your code logic. I think there are two reasonable options.

1. add the validation logic (regex or %in% or between(), etc.) directly in SURVEY.R as an argument:
```r
sd_question(
  type = "text",
  id = "mail",
  label = "Please provide your email adress:",
  validate = c(
    "No valid email address." = "^.+?@.+?\..{2,}$"
  )
```
or 2. add a new helper function similar to sd_skip_if() in the APP.R server function:
```r
sd_validate(
  !grepl("^.+?@.+?\..{2,}$", input$mail) ~ "No valid email address"
)
```

One additional point I wasn't able to solve with the shinyvalidate package, but what could be helpful in a new implementation, the comparison of the values from two inputs. For example, if you let the respondents input their email address twice:

1. version in SURVEY.R:
```r
sd_question(
  type = "text",
  id = "mail2",
  label = "Please provide your email adress:",
  validate = c(
    "email addresses are not the same" = mail != mail2
  )
```
or 2. version APP.R server function:
```r
sd_validate(
  input$mail != input$mail2 ~ "email addresses are not the same" # but unclear, where to output this error message
)
```

**Comment by jhelvy** [comment]
Cool, yes input validation is on our roadmap. I like the suggestion of adding a `sd_validate()` function to add validation conditions. We'll look into this.

**Comment by jhelvy** [comment]
FYI, this is an older discussion thread, but we now have `sd_stop_if()` for validating question responses before being allowed to continue in the survey. It isn't quite the same as this discussion, but it can be used to achieve the same goal. Here's an update on the related issue: https://github.com/surveydown-dev/surveydown/issues/216#issuecomment-3286139523
